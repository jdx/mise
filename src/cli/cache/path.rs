use eyre::Result;

use crate::env;

/// Show the cache directory path
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, visible_alias = "dir")]
pub struct CachePath {}

impl CachePath {
    pub fn run(self) -> Result<()> {
        miseprintln!("{}", env::MISE_CACHE_DIR.display());
        Ok(())
    }
}
