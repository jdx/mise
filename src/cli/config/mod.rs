use clap::Subcommand;
use eyre::Result;

mod generate;
mod ls;

/// [experimental] Manage config files
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "cfg")]
pub struct Config {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    no_header: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Ls(ls::ConfigLs),
    Generate(generate::ConfigGenerate),
}

impl Commands {
    pub fn run(self) -> Result<()> {
        match self {
            Self::Ls(cmd) => cmd.run(),
            Self::Generate(cmd) => cmd.run(),
        }
    }
}

impl Config {
    pub fn run(self) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::ConfigLs {
            no_header: self.no_header,
        }));

        cmd.run()
    }
}
