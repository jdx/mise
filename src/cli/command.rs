use color_eyre::eyre::Result;

use crate::config::Config;
use crate::output::Output;

/// described a CLI command
///
/// e.g.: `rtx plugins ls`
pub trait Command: Sized {
    /// CLI command entry point
    fn run(self, config: Config, output: &mut Output) -> Result<()>;
}
