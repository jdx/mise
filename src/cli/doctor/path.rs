use crate::config::Config;
use crate::Result;
use std::env;

/// Print the current PATH entries mise is providing
#[derive(Debug, clap::Args)]
#[clap(alias="paths", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Path {
    /// Print all entries including those not provided by mise
    #[clap(long, short, verbatim_doc_comment)]
    full: bool,
}

impl Path {
    pub fn run(self) -> Result<()> {
        let config = Config::get();
        let ts = config.get_toolset()?;
        let paths = if self.full {
            let env = ts.env_with_path(&config)?;
            let path = env.get("PATH").cloned().unwrap_or_default();
            env::split_paths(&path).collect()
        } else {
            ts.list_final_paths()?
        };
        for path in paths {
            println!("{}", path.display());
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    Get the current PATH entries mise is providing
    $ mise path
    /home/user/.local/share/mise/installs/node/24.0.0/bin
    /home/user/.local/share/mise/installs/rust/1.90.0/bin
    /home/user/.local/share/mise/installs/python/3.10.0/bin
"#
);
