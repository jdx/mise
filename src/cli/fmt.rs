use crate::config::ALL_TOML_CONFIG_FILES;
use crate::Result;
use crate::{config, dirs, file};
use eyre::bail;
use taplo::formatter::Options;

/// Formats mise.toml
///
/// Sorts keys and cleans up whitespace in mise.toml
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Fmt {
    /// Format all files from the current directory
    #[clap(short, long)]
    pub all: bool,
}

impl Fmt {
    pub fn run(self) -> eyre::Result<()> {
        let cwd = dirs::CWD.clone().unwrap_or_default();
        let configs = if self.all {
            ALL_TOML_CONFIG_FILES.clone()
        } else {
            config::config_files_in_dir(&cwd)
        };
        if configs.is_empty() {
            bail!("No config file found in current directory");
        }
        for p in configs {
            if !p
                .file_name()
                .is_some_and(|f| f.to_string_lossy().ends_with("toml"))
            {
                continue;
            }
            let toml = file::read_to_string(&p)?;
            let toml = sort(toml)?;
            let toml = taplo::formatter::format(
                &toml,
                Options {
                    align_entries: false,
                    align_comments: true,
                    align_single_comments: true,
                    array_trailing_comma: true,
                    array_auto_expand: true,
                    inline_table_expand: true,
                    array_auto_collapse: true,
                    compact_arrays: true,
                    compact_inline_tables: false,
                    compact_entries: false,
                    column_width: 80,
                    indent_tables: false,
                    indent_entries: false,
                    indent_string: "  ".to_string(),
                    trailing_newline: true,
                    reorder_keys: false,
                    reorder_arrays: false,
                    allowed_blank_lines: 2,
                    crlf: false,
                },
            );
            file::write(&p, &toml)?;
        }

        Ok(())
    }
}

fn sort(toml: String) -> Result<String> {
    let mut doc: toml_edit::DocumentMut = toml.parse()?;
    let order = |k: String| match k.as_str() {
        "min_version" => 0,
        "env_file" => 1,
        "env_path" => 2,
        "_" => 3,
        "env" => 4,
        "vars" => 5,
        "hooks" => 6,
        "watch_files" => 7,
        "tools" => 8,
        "tasks" => 10,
        "task_config" => 11,
        "redactions" => 12,
        "alias" => 13,
        "plugins" => 14,
        "settings" => 15,
        _ => 9,
    };
    doc.sort_values_by(|a, _, b, _| order(a.to_string()).cmp(&order(b.to_string())));
    Ok(doc.to_string())
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    $ <bold>mise fmt</bold>
"#
);
