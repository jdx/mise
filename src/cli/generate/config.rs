use crate::Result;
use crate::cli::config::generate;

/// [experimental] Generate a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = generate::AFTER_LONG_HELP)]
pub struct Config {
    #[clap(flatten)]
    generate: generate::ConfigGenerate,
}

impl Config {
    pub async fn run(self) -> Result<()> {
        self.generate.run()
    }
}
