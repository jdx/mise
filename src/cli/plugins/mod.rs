use clap::Subcommand;
use eyre::Result;

use crate::config::Config;

mod install;
mod link;
mod ls;
mod ls_remote;
mod uninstall;
mod update;

#[derive(Debug, clap::Args)]
#[clap(about = "Manage plugins", visible_alias = "p", aliases = ["plugin", "plugin-list"])]
pub struct Plugins {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// list all available remote plugins
    ///
    /// same as `rtx plugins ls-remote`
    #[clap(short, long, hide = true)]
    pub all: bool,

    /// The built-in plugins only
    /// Normally these are not shown
    #[clap(short, long, verbatim_doc_comment, conflicts_with = "all")]
    pub core: bool,

    /// List installed plugins
    ///
    /// This is the default behavior but can be used with --core
    /// to show core and user plugins
    #[clap(long, verbatim_doc_comment, conflicts_with = "all")]
    pub user: bool,

    /// show the git url for each plugin
    ///
    /// e.g.: https://github.com/asdf-vm/asdf-node.git
    #[clap(short, long)]
    pub urls: bool,

    /// show the git refs for each plugin
    ///
    /// e.g.: main 1234abc
    #[clap(long)]
    pub refs: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Install(install::PluginsInstall),
    Link(link::PluginsLink),
    Ls(ls::PluginsLs),
    LsRemote(ls_remote::PluginsLsRemote),
    Uninstall(uninstall::PluginsUninstall),
    Update(update::Update),
}

impl Commands {
    pub fn run(self, config: &Config) -> Result<()> {
        match self {
            Self::Install(cmd) => cmd.run(config),
            Self::Link(cmd) => cmd.run(),
            Self::Ls(cmd) => cmd.run(config),
            Self::LsRemote(cmd) => cmd.run(config),
            Self::Uninstall(cmd) => cmd.run(config),
            Self::Update(cmd) => cmd.run(config),
        }
    }
}

impl Plugins {
    pub fn run(self, config: &Config) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::PluginsLs {
            all: self.all,
            core: self.core,
            refs: self.refs,
            urls: self.urls,
            user: self.user,
        }));

        cmd.run(config)
    }
}
