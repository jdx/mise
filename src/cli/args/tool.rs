use std::ffi::{OsStr, OsString};
use std::fmt::Display;

use clap::{Arg, Command, Error};
use color_eyre::eyre::Result;
use regex::Regex;

use crate::plugins::PluginName;
use crate::toolset::ToolVersionRequest;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ToolArg {
    pub plugin: PluginName,
    pub tvr: Option<ToolVersionRequest>,
}

impl ToolArg {
    pub fn parse(input: &str) -> Self {
        match input.split_once('@') {
            Some((plugin, version)) => Self {
                plugin: plugin.to_string(),
                tvr: Some(ToolVersionRequest::new(plugin.to_string(), version)),
            },
            None => Self {
                plugin: input.into(),
                tvr: None,
            },
        }
    }

    /// this handles the case where the user typed in:
    /// rtx local node 20.0.0
    /// instead of
    /// rtx local node@20.0.0
    ///
    /// We can detect this, and we know what they meant, so make it work the way
    /// they expected.
    pub fn double_tool_condition(tools: &[ToolArg]) -> Vec<ToolArg> {
        let mut tools = tools.to_vec();
        if tools.len() == 2 {
            let re: &Regex = regex!(r"^\d+(\.\d+)?(\.\d+)?$");
            let a = tools[0].clone();
            let b = tools[1].clone();
            if matches!(a.tvr, None) && matches!(b.tvr, None) && re.is_match(&b.plugin) {
                tools[1].tvr = Some(ToolVersionRequest::new(a.plugin.clone(), &b.plugin));
                tools[1].plugin = a.plugin;
                tools.remove(0);
            }
        }
        tools
    }

    pub fn with_version(self, version: &str) -> Self {
        Self {
            tvr: Some(ToolVersionRequest::new(self.plugin.clone(), version)),
            ..self
        }
    }
}

impl Display for ToolArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.tvr {
            Some(tvr) => write!(f, "{}", tvr),
            _ => write!(f, "{}", self.plugin),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ToolArgParser;

impl clap::builder::TypedValueParser for ToolArgParser {
    type Value = ToolArg;

    fn parse_ref(
        &self,
        cmd: &Command,
        arg: Option<&Arg>,
        value: &OsStr,
    ) -> Result<Self::Value, Error> {
        self.parse(cmd, arg, value.to_os_string())
    }

    fn parse(
        &self,
        _cmd: &Command,
        _arg: Option<&Arg>,
        value: OsString,
    ) -> Result<Self::Value, Error> {
        Ok(ToolArg::parse(&value.to_string_lossy()))
    }
}
