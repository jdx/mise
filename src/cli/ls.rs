use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::path::PathBuf;
use std::sync::Arc;

use color_eyre::eyre::Result;
use console::style;
use console::Alignment::Left;
use indexmap::IndexMap;
use itertools::Itertools;
use serde_derive::Serialize;
use versions::Versioning;

use crate::cli::command::Command;
use crate::config::Config;
use crate::errors::Error::PluginNotInstalled;
use crate::output::Output;
use crate::plugins::PluginName;
use crate::tool::Tool;
use crate::toolset::{ToolSource, ToolVersion, ToolsetBuilder};

/// List installed and/or currently selected tool versions
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "list", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Ls {
    /// Only show tool versions from [PLUGIN]
    #[clap(long, short)]
    plugin: Option<PluginName>,

    #[clap(hide = true)]
    plugin_arg: Option<PluginName>,

    /// Only show tool versions currently specified in a .tool-versions/.rtx.toml
    #[clap(long, short)]
    current: bool,

    /// Only show tool versions that are installed
    /// Hides missing ones defined in .tool-versions/.rtx.toml but not yet installed
    #[clap(long, short)]
    installed: bool,

    /// Output in an easily parseable format
    #[clap(long, hide = true, visible_short_alias = 'x', conflicts_with = "json")]
    parseable: bool,

    /// Output in json format
    #[clap(long, visible_short_alias = 'J', overrides_with = "parseable")]
    json: bool,

    /// Display missing tool versions
    #[clap(long, short, conflicts_with = "installed")]
    missing: bool,

    /// Display versions matching this prefix
    #[clap(long)]
    prefix: Option<String>,
}

impl Command for Ls {
    fn run(mut self, mut config: Config, out: &mut Output) -> Result<()> {
        self.plugin = self.plugin.clone().or(self.plugin_arg.clone());
        self.verify_plugin(&config)?;

        let mut runtimes = self.get_runtime_list(&mut config)?;
        if self.current {
            runtimes.retain(|(_, _, source)| source.is_some());
        }
        if self.installed {
            runtimes.retain(|(p, tv, _)| p.is_version_installed(tv));
        }
        if self.missing {
            runtimes.retain(|(p, tv, _)| !p.is_version_installed(tv));
        }
        if let Some(prefix) = &self.prefix {
            if self.plugin.is_none() {
                panic!("--prefix requires --plugin");
            }
            runtimes.retain(|(_, tv, _)| tv.version.starts_with(prefix));
        }
        if self.json {
            self.display_json(runtimes, out)
        } else if self.parseable {
            self.display_parseable(runtimes, out)
        } else {
            self.display_user(&config, runtimes, out)
        }
    }
}

type JSONOutput = IndexMap<PluginName, Vec<JSONToolVersion>>;

#[derive(Serialize)]
struct JSONToolVersion {
    version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    requested_version: Option<String>,
    install_path: PathBuf,
    #[serde(skip_serializing_if = "Option::is_none")]
    source: Option<IndexMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    symlinked_to: Option<PathBuf>,
}

impl Ls {
    fn verify_plugin(&self, config: &Config) -> Result<()> {
        match &self.plugin {
            Some(plugin_name) => {
                let plugin = config.tools.get(plugin_name);
                if plugin.is_none() || !plugin.unwrap().is_installed() {
                    return Err(PluginNotInstalled(plugin_name.clone()))?;
                }
            }
            None => {}
        }
        Ok(())
    }

    fn display_json(&self, runtimes: Vec<RuntimeRow>, out: &mut Output) -> Result<()> {
        let mut plugins = JSONOutput::new();
        for (plugin_name, runtimes) in &runtimes
            .into_iter()
            .group_by(|(p, _, _)| p.name.to_string())
        {
            let runtimes = runtimes
                .map(|(p, tv, source)| JSONToolVersion {
                    symlinked_to: p.symlink_path(&tv),
                    install_path: tv.install_path(),
                    version: tv.version,
                    requested_version: source.as_ref().map(|_| tv.request.version()),
                    source: source.map(|source| source.as_json()),
                })
                .collect();
            if self.plugin.is_some() {
                // only display 1 plugin
                out.stdout.writeln(serde_json::to_string_pretty(&runtimes)?);
                return Ok(());
            }
            plugins.insert(plugin_name.clone(), runtimes);
        }
        out.stdout.writeln(serde_json::to_string_pretty(&plugins)?);
        Ok(())
    }

    fn display_parseable(&self, runtimes: Vec<RuntimeRow>, out: &mut Output) -> Result<()> {
        warn!("The parseable output format is deprecated and will be removed in a future release.");
        warn!("Please use the regular output format instead which has been modified to be more easily parseable.");
        runtimes
            .into_iter()
            .map(|(p, tv, _)| (p, tv))
            .filter(|(p, tv)| p.is_version_installed(tv))
            .for_each(|(_, tv)| {
                if self.plugin.is_some() {
                    // only displaying 1 plugin so only show the version
                    rtxprintln!(out, "{}", tv.version);
                } else {
                    rtxprintln!(out, "{} {}", tv.plugin_name, tv.version);
                }
            });
        Ok(())
    }

    fn display_user(
        &self,
        config: &Config,
        runtimes: Vec<RuntimeRow>,
        out: &mut Output,
    ) -> Result<()> {
        let output = runtimes
            .into_iter()
            .map(|(p, tv, source)| {
                let plugin = p.name.to_string();
                let version = if let Some(symlink_path) = p.symlink_path(&tv) {
                    VersionStatus::Symlink(tv.version, symlink_path)
                } else if !p.is_version_installed(&tv) {
                    VersionStatus::Missing(tv.version)
                } else if source.is_some() {
                    VersionStatus::Active(tv.version.clone(), p.is_version_outdated(config, &tv))
                } else {
                    VersionStatus::Inactive(tv.version)
                };
                let request = source.map(|source| (source.to_string(), tv.request.version()));
                (plugin, version, request)
            })
            .collect::<Vec<_>>();
        let (max_plugin_len, max_version_len, max_source_len) = output.iter().fold(
            (0, 0, 0),
            |(max_plugin, max_version, max_source), (plugin, version, request)| {
                let plugin = max_plugin.max(plugin.len());
                let version = max_version.max(version.to_plain_string().len());
                let source = match request {
                    Some((source, _)) => max_source.max(source.len()),
                    None => max_source,
                };
                (plugin.min(10), version.min(15), source.min(30))
            },
        );
        for (plugin, version, request) in output {
            let pad = |s, len| console::pad_str(s, len, Left, None);
            let plugin_extra =
                ((plugin.len() as i8 - max_plugin_len as i8).max(0) as usize).min(max_version_len);
            let plugin = pad(&plugin, max_plugin_len);
            let plugin = style(plugin).cyan();
            let version_extra = (version.to_plain_string().len() as i8 - max_version_len as i8
                + plugin_extra as i8)
                .max(0) as usize;
            let version = version.to_string();
            let version = pad(&version, max_version_len - plugin_extra);
            let line = match &request {
                Some((source, requested)) => {
                    let source = pad(source, max_source_len - version_extra);
                    format!("{} {} {} {}", plugin, version, source, requested)
                }
                None => {
                    format!("{} {}", plugin, version)
                }
            };
            rtxprintln!(out, "{}", line.trim_end());
        }
        Ok(())
    }

    fn get_runtime_list(&self, config: &mut Config) -> Result<Vec<RuntimeRow>> {
        let ts = ToolsetBuilder::new().build(config)?;
        let mut versions: HashMap<(PluginName, String), (Arc<Tool>, ToolVersion)> = ts
            .list_installed_versions(config)?
            .into_iter()
            .filter(|(p, _)| match &self.plugin {
                Some(plugin) => &p.name == plugin,
                None => true,
            })
            .map(|(p, tv)| ((p.name.clone(), tv.version.clone()), (p, tv)))
            .collect();

        let active = ts
            .list_current_versions(config)
            .into_iter()
            .map(|(p, tv)| ((p.name.clone(), tv.version.clone()), (p, tv)))
            .collect::<HashMap<(PluginName, String), (Arc<Tool>, ToolVersion)>>();

        versions.extend(
            active
                .clone()
                .into_iter()
                .filter(|((plugin_name, _), _)| match &self.plugin {
                    Some(plugin) => plugin_name == plugin,
                    None => true,
                })
                .collect::<Vec<((PluginName, String), (Arc<Tool>, ToolVersion))>>(),
        );

        let rvs: Vec<RuntimeRow> = versions
            .into_iter()
            .sorted_by_cached_key(|((plugin_name, version), _)| {
                (plugin_name.clone(), Versioning::new(version))
            })
            .map(|(k, (p, tv))| {
                let source = match &active.get(&k) {
                    Some((_, tv)) => ts.versions.get(&tv.plugin_name).map(|tv| tv.source.clone()),
                    None => None,
                };
                (p, tv, source)
            })
            // if it isn't installed and it's not specified, don't show it
            .filter(|(p, tv, source)| source.is_some() || p.is_version_installed(tv))
            .collect();

        Ok(rvs)
    }
}

type RuntimeRow = (Arc<Tool>, ToolVersion, Option<ToolSource>);

enum VersionStatus {
    Active(String, bool),
    Inactive(String),
    Missing(String),
    Symlink(String, PathBuf),
}

impl VersionStatus {
    fn to_plain_string(&self) -> String {
        match self {
            VersionStatus::Active(version, outdated) => {
                if *outdated {
                    format!("{} (outdated)", version)
                } else {
                    version.to_string()
                }
            }
            VersionStatus::Inactive(version) => version.to_string(),
            VersionStatus::Missing(version) => format!("{} (missing)", version),
            VersionStatus::Symlink(version, _) => format!("{} (symlink)", version),
        }
    }
}

impl Display for VersionStatus {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            VersionStatus::Active(version, outdated) => {
                if *outdated {
                    write!(
                        f,
                        "{} {}",
                        style(version).yellow(),
                        style("(outdated)").yellow()
                    )
                } else {
                    write!(f, "{}", style(version).green())
                }
            }
            VersionStatus::Inactive(version) => write!(f, "{}", style(version).dim()),
            VersionStatus::Missing(version) => write!(
                f,
                "{} {}",
                if console::colors_enabled() {
                    style(version).strikethrough().red().to_string()
                } else {
                    version.to_string()
                },
                style("(missing)").red()
            ),
            VersionStatus::Symlink(version, _) => {
                write!(f, "{} {}", version, style("(symlink)").dim())
            }
        }
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  $ <bold>rtx ls</bold>
  node    20.0.0 ~/src/myapp/.tool-versions latest
  python  3.11.0 ~/.tool-versions           3.10
  python  3.10.0

  $ <bold>rtx ls --current</bold>
  node    20.0.0 ~/src/myapp/.tool-versions 20
  python  3.11.0 ~/.tool-versions           3.11.0

  $ <bold>rtx ls --json</bold>
  {
    "node": [
      {
        "version": "20.0.0",
        "install_path": "/Users/jdx/.rtx/installs/node/20.0.0",
        "source": {
          "type": ".rtx.toml",
          "path": "/Users/jdx/.rtx.toml"
        }
      }
    ],
    "python": [...]
  }
"#
);

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_str_eq;

    use crate::file::remove_all;
    use crate::{assert_cli, assert_cli_err, assert_cli_snapshot, dirs};

    #[test]
    fn test_ls() {
        let _ = remove_all(dirs::INSTALLS.as_path());
        assert_cli!("install");
        assert_cli_snapshot!("list");

        assert_cli!("install", "tiny@2.0.0");
        assert_cli_snapshot!("list");

        assert_cli!("uninstall", "tiny@3.1.0");
        assert_cli_snapshot!("list");

        assert_cli!("uninstall", "tiny@2.0.0");
        assert_cli_snapshot!("list");

        assert_cli!("install");
        assert_cli_snapshot!("list");
    }

    #[test]
    fn test_ls_current() {
        assert_cli_snapshot!("ls", "-c");
    }

    #[test]
    fn test_ls_json() {
        let _ = remove_all(dirs::INSTALLS.as_path());
        assert_cli!("install");
        assert_cli_snapshot!("ls", "--json");
        assert_cli_snapshot!("ls", "--json", "tiny");
    }

    #[test]
    fn test_ls_parseable() {
        let _ = remove_all(dirs::INSTALLS.as_path());
        assert_cli!("install");
        assert_cli_snapshot!("ls", "-x");
        assert_cli_snapshot!("ls", "--parseable", "tiny");
    }

    #[test]
    fn test_ls_missing() {
        assert_cli!("install");
        assert_cli_snapshot!("ls", "--missing");
    }

    #[test]
    fn test_ls_missing_plugin() {
        let err = assert_cli_err!("ls", "missing-plugin");
        assert_str_eq!(err.to_string(), r#"[missing-plugin] plugin not installed"#);
    }

    #[test]
    fn test_ls_prefix() {
        assert_cli!("install");
        assert_cli_snapshot!("ls", "--plugin=tiny", "--prefix=3");
    }
}
