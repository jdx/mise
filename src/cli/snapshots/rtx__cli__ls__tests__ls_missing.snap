---
source: src/cli/ls.rs
expression: output
---

